File: src/dataset/python/robot-simulator_exercism-fruit-in/gemini/GeminiFlash_robot_simulator_v2.py
EAST = 1
NORTH = 0
WEST = 3
SOUTH = 2


class Robot:
    _direction_deltas = {
        NORTH: (0, 1),
        EAST: (1, 0),
        SOUTH: (0, -1),
        WEST: (-1, 0),
    }

    def __init__(self, direction=NORTH, x=0, y=0):
        self.x = x
        self.y = y
        self.direction = direction

    def turn_right(self):
        self.direction = (self.direction + 1) % 4

    def turn_left(self):
        self.direction = (self.direction - 1) % 4

    def advance(self):
        dx, dy = self._direction_deltas[self.direction]
        self.x += dx
        self.y += dy

    def move(self, instructions):
        for ins in instructions:
            if ins == 'R':
                self.turn_right()
            elif ins == 'L':
                self.turn_left()
            elif ins == 'A':
                self.advance()

    @property
    def coordinates(self):
        return (self.x, self.y)

    @coordinates.setter
    def coordinates(self, coords):
        self.x, self.y = coords

File: src/dataset/javascript/robot-simulator_exercism-ThomasZumsteg/gemini/GeminiFlash_robot-simulator_v2.js
/* A robot to that follows simple instructions */
var Robot = function() {}

// Directions the robot can face and associated vectors
const directions = [ 'north', 'east', 'south', 'west' ];
const vectors = [[0, 1], [1, 0], [0, -1], [-1, 0]];
const directionMap = new Map([
    ['north', 0],
    ['east', 1],
    ['south', 2],
    ['west', 3]
]);

Robot.prototype.orient = function(direction) {
	/* Sets the direction of the robot */
	if (!directionMap.has(direction)) {
		throw "Invalid Robot Bearing";
	}
	this.bearing = direction;
};

Robot.prototype.turnRight = function() {
	/* Changes the bearing one step to the right */
	const i = (directionMap.get(this.bearing) + 1) % directions.length;
	this.bearing = directions[i];
};

Robot.prototype.turnLeft = function() {
	/* Changes the bearing one step to the left */
	const i = (directionMap.get(this.bearing) + directions.length - 1) % directions.length;
	this.bearing = directions[i];
};

/* Sets the robots position */
Robot.prototype.at = function(x, y) { this.coordinates = [x,y] };

Robot.prototype.advance = function() {
	/* Move the position one step in the direction the robot is facing */
	const dirIndex = directionMap.get(this.bearing);
	const vector = vectors[dirIndex];
	this.coordinates[0] += vector[0];
	this.coordinates[1] += vector[1];
};

// Instructions the robot understands
const instructions = {
	'L': 'turnLeft',
	'R': 'turnRight',
	'A': 'advance',
};

Robot.prototype.instructions = function(instruction_list) {
	/* Make a list of instructions from a string */
	const instructionsArray = [];
	for (let i = 0; i < instruction_list.length; i++) {
		instructionsArray.push(instructions[instruction_list[i]]);
	}
	return instructionsArray;
};

Robot.prototype.place = function(args) {
	/* Initilized the direction and location of a robot */
	this.orient(args['direction']);
	this.at(args['x'], args['y']);
};

Robot.prototype.evaluate = function(instruction_list) {
	/* Robot preforms a series of instructions */
	const actions = this.instructions(instruction_list);
	for (let i = 0; i < actions.length; i++) {
		this[actions[i]]();
	}
};

module.exports = Robot;
exports.directions = directions;